### ISSUE TITLE
NO_PUBKEY 5BB92C09DB82666C

### ISSUE THREAD

**Issue by wsfulton** [deadsnakes/issues#53]
### description
Has the repository key changed recently? We use deadsnakes on Travis CI but I can replicate the problem on a local box as shown below:
```
$ sudo add-apt-repository -y ppa:fkrull/deadsnakes
gpg: keyring `/tmp/tmpalqa6856/secring.gpg' created
gpg: keyring `/tmp/tmpalqa6856/pubring.gpg' created
gpg: "tag:launchpad.net:2008:redacted" not a key ID: skipping
$ sudo apt-get -qq update
W: GPG error: http://ppa.launchpad.net trusty InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 5BB92C09DB82666C
```
Adding the following before hand seems to get it to work again:
```
$ sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 5BB92C09DB82666C
```
but seems more like a workaround than a fix. Is this something that should 'just work' like it used to?
#### lsb_release -a

```
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 14.04.5 LTS
Release:	14.04
Codename:	trusty
```

#### uname -a

```
Linux ubuntu 4.4.0-104-generic #127~14.04.1-Ubuntu SMP Mon Dec 11 12:44:15 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by asottile** [comment]
(not at a computer so I can't definitively check)

We recently moved ppa names, can you try ppa:deadsnakes instead?

**Comment by wsfulton** [comment]
Ah, this does 'just work':
```
sudo add-apt-repository -y ppa:deadsnakes/ppa
```
So it seems that there is no backwards compatibility with the old ppa name, and one simply has to use a differently named ppa. That's fine, it just wasn't clearly stated anywhere I could find that `ppa:fkrull/deadsnakes` should be replaced by `ppa:deadsnakes/ppa`. A rather subtle change!

**Comment by asottile** [comment]
Ah yeah, for several (6?) months we had this on the homepage of fkrull/deadsnakes:

> Please use the new repository at
>
> https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa
>
> instead!

and it looks like we finally closed the door in https://github.com/deadsnakes/issues/issues/46#issuecomment-353780893

I wasn't aware this would mess with the public keys though, CC @fkrull

**Comment by fkrull** [comment]
Well, when you disable a PPA, the packages are still there, but Launchpad does something so you can't get the public key for it any more. That means old "installations" of the repository still work, but you can't add it to a new system (well, you could copy over the key or something silly).

Launchpad doesn't have a way to deprecate or move repositories. All you can do is put a note in the description, but people who don't visit the PPA page have no way to know. It's quite frustrating, really. So I figured waiting any longer isn't going to help much. Sorry, but now you know. :)

**Comment by sonakhan** [comment]
The reason why I added deadsnakes was to get python 3.6, but that stops working with the new ppa:deadsnakes/ppa change. Help?

**Comment by asottile** [comment]
@sonakhan can you provide some steps to reproduce?  I'm using python3.6 just fine with the following:

(Here's a sample dockerfile)

```dockerfile
FROM ubuntu:xenial
RUN : \
    && apt-get update \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
        software-properties-common \
    && add-apt-repository ppa:deadsnakes \
    && apt-get update \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
        python3.6-dev \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*
```

```console
$ docker build -t test .

...

$ docker run -ti test python3.6
Python 3.6.4 (default, Jan 28 2018, 17:52:01) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> thing = 'world'
>>> f'hello {thing}'
'hello world'
```

I'm also using it on my desktop installation:

```console
$ cat /etc/apt/sources.list.d/deadsnakes-ubuntu-ppa-xenial.list 
deb http://ppa.launchpad.net/deadsnakes/ppa/ubuntu xenial main
# deb-src http://ppa.launchpad.net/deadsnakes/ppa/ubuntu xenial main
$ dpkg -l | grep python3.6
ii  libpython3.6:amd64                          3.6.4-1+xenial2                              amd64        Shared Python runtime library (version 3.6)
ii  libpython3.6-dev:amd64                      3.6.4-1+xenial2                              amd64        Header files and a static library for Python (v3.6)
ii  libpython3.6-minimal:amd64                  3.6.4-1+xenial2                              amd64        Minimal subset of the Python language (version 3.6)
ii  libpython3.6-stdlib:amd64                   3.6.4-1+xenial2                              amd64        Interactive high-level object-oriented language (standard library, version 3.6)
ii  libpython3.6-testsuite                      3.6.4-1+xenial2                              all          Testsuite for the Python standard library (v3.6)
ii  python3.6                                   3.6.4-1+xenial2                              amd64        Interactive high-level object-oriented language (version 3.6)
ii  python3.6-dev                               3.6.4-1+xenial2                              amd64        Header files and a static library for Python (v3.6)
ii  python3.6-minimal                           3.6.4-1+xenial2                              amd64        Minimal subset of the Python language (version 3.6)
```
```

**Comment by sonakhan** [comment]
Here is my dockerfile: 

FROM ubuntu:16.04

RUN apt-get upgrade --fix-missing
RUN apt-get update
RUN apt-get install -y --no-install-recommends software-properties-common
RUN apt-get install -y build-essential
RUN apt-get install -y zip
RUN apt-get install -y wget
RUN apt-get install -y curl
RUN add-apt-repository ppa:deadsnakes/ppa

RUN apt-get install -y python3.6

and it gets stuck here: 

Step 10/23 : RUN apt-get install -y python3.6
 ---> Running in 032d1158e01b
Reading package lists...
Building dependency tree...
Reading state information...
E: Unable to locate package python3.6
E: Couldn't find any package by glob 'python3.6'
E: Couldn't find any package by regex 'python3.6'
The command '/bin/sh -c apt-get install -y python3.6' returned a non-zero code: 100

**Comment by sonakhan** [comment]
Doing 

`RUN add-apt-repository ppa:deadsnakes
`
gives similar error (as with ".../ppa")

**Comment by asottile** [comment]
You need an `apt-get update` after adding / removing PPAs to be able to install from them

Also from [docker best practices](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#apt-get), you'll want to combine your `apt` install lines into one (as I did in my example `Dockerfile` above)

**Comment by sonakhan** [comment]
Thank you so much, @asottile . That worked :) 
Could you explain why another apt-get update is needed?

**Comment by asottile** [comment]
Sure!

Running `apt-get update` populates a local cache of the remote package servers.  The first `apt-get update` populates a local cache of whatever is in your `sources.list` at _that time_ (in your case, it'll just be the default ubuntu sources).

Adding a PPA adds a new source of packages, but the remote metadata hasn't been pulled in since adding the PPA.  An `apt-get update` after that will pull in the metadata from the additional repository.

Basically, when you add a PPA, you're saying "next time I `apt-get update` also fetch the metadata from these sources" and without an `apt-get update`, the metadata for those additional packages isn't available.

Some other package managers don't have this two-step metadata transfer, for instance `pip` / `npm` always reach out to the package sources for metadata requests.

**Comment by jenstroeger** [comment]
I’d like to add the deadsnakes PPA manually to `focal`:
```
> echo -e "deb http://ppa.launchpad.net/deadsnakes/ppa/ubuntu focal main\n# deb-src http://ppa.launchpad.net/deadsnakes/ppa/ubuntu focal main" > /etc/apt/sources.list.d/deadsnakes-ubuntu-ppa-focal.list
```
and need the public key. However, I can’t find the key’s URL documented?

**Comment by asottile** [comment]
`add-apt-repository ppa:deadsnakes` will add it for you.  there isn't a directly downloadable url since the keys are managed by launchpad

one thing you can do is retrieve it from ubuntu's keyservers though: `apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys F23C5A6CF475977595C89F51BA6932366A755776`

but you're probably better to just use `add-apt-repository` which is the supported and future-proof approach

**Comment by asottile** [comment]
I'm going to lock this issue since all comments on it have been off topic from the original issue -- if you have a new problem please open a new issue
